client/client.go

```diff
@@ -122,25 +123,41 @@ func bind(cfg *Config, conn ldaputil.Connection) error {
 		return errors.New("unable to bind due to lack of configured password")
 	}
 
-	if cfg.BindDN == "" {
-		return errors.New("must provide binddn")
-	}
-
-	origErr := conn.Bind(cfg.BindDN, cfg.BindPassword)
-	if origErr == nil {
+	if cfg.UPNDomain != "" {
+		origErr := conn.Bind(fmt.Sprintf("%s@%s", ldaputil.EscapeLDAPValue(cfg.BindDN), cfg.UPNDomain), cfg.BindPassword)
+		if origErr == nil {
+			return nil
+		}
+		if !shouldTryLastPwd(cfg.LastBindPassword, cfg.LastBindPasswordRotation) {
+			return origErr
+		}
+		if err := conn.Bind(fmt.Sprintf("%s@%s", ldaputil.EscapeLDAPValue(cfg.BindDN), cfg.UPNDomain), cfg.LastBindPassword); err != nil {
+			// Return the original error because it'll be more helpful for debugging.
+			return origErr
+		}
 		return nil
 	}
-	if !shouldTryLastPwd(cfg.LastBindPassword, cfg.LastBindPasswordRotation) {
-		return origErr
-	}
 
-	return conn.Bind(cfg.BindDN, cfg.LastBindPassword)
+	if cfg.BindDN != "" {
+		origErr := conn.Bind(cfg.BindDN, cfg.BindPassword)
+		if origErr == nil {
+			return nil
+		}
+		if !shouldTryLastPwd(cfg.LastBindPassword, cfg.LastBindPasswordRotation) {
+			return origErr
+		}
+		if err := conn.Bind(cfg.BindDN, cfg.LastBindPassword); err != nil {
```

Comment:
I think if this call succeeds, we'll fall through to the `"must provide binddn or upndomain"` error. Using a switch-case (or an if-else chain, but then I think lots of Go linters will complain about unnecessary indentation) instead of if statements would make it clearer that we expect to take precisely 1 of the 3 code paths, and easier to avoid that bug.

Reply:
This bug also exists in AD secrets ([client.go#L143-L156](https://github.com/hashicorp/vault-plugin-secrets-ad/blob/main/plugin/client/client.go#L143-L156)). I'll get it fixed up in both places.

Reply:
Looking closer at AD, this bug (fixed here in https://github.com/hashicorp/vault-plugin-secrets-openldap/pull/42) also exists. Will also get that fixed in AD secrets.

Reply:
I've refactored this to fix the bug in https://github.com/hashicorp/vault-plugin-secrets-openldap/commit/d7d27a848321e1d1fd77395453c338546715a0fa. Thanks!